File: LeetCode/39.组合总和/39.组合总和.cpp
class Solution {
public:
    void dfs(vector<int>& candidates, vector<vector<int>> &ans, vector<int> &comb,int target, int index){
        int n = candidates.size();
        if(index >= n){
            return;
        }
        if(target == 0){
            ans.emplace_back(comb);
            return;
        }
        dfs(candidates, ans, comb, target, index + 1);
        if(target - candidates[index] >= 0){
            comb.emplace_back(candidates[index]);
            dfs(candidates, ans, comb, target - candidates[index], index);
            comb.pop_back();
        }

    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> comb;
        dfs(candidates, ans, comb, target, 0);
        return ans;
    }
};

File: LeetCode/39.组合总和/39.组合总和.go
func combinationSum(candidates []int, target int) (ans [][]int) {
    comb := []int{}
    var dfs func (target, index int)
    
    dfs = func(target, index int){
        if index == len(candidates){
            return
        }
        if target == 0 {
            ans = append(ans, append([]int(nil), comb...))
            return
        }
        dfs(target, index + 1)
        if target - candidates[index] >= 0 {
            comb = append(comb, candidates[index])
            dfs(target - candidates[index], index)
            comb = comb[:len(comb) -1]
        }
    }
    dfs(target, 0)
    return
    
}